lib/web_ui/lib/src/engine/html/shaders/shader.dart

```diff
@@ -218,30 +350,95 @@ class GradientRadial extends EngineGradient {
+      return _createCanvasGradient(ctx, shaderBounds, density);
+    } else {
+      initWebGl();
+      return _createGlGradient(ctx, shaderBounds, density);
     }
+  }
+
+  Object _createCanvasGradient(html.CanvasRenderingContext2D? ctx,
+      ui.Rect? shaderBounds, double density) {
     final double offsetX = shaderBounds!.left;
     final double offsetY = shaderBounds.top;
     final html.CanvasGradient gradient = ctx!.createRadialGradient(
         center.dx - offsetX, center.dy - offsetY, 0,
         center.dx - offsetX, center.dy - offsetY, radius);
-    final List<double>? colorStops = this.colorStops;
-    if (colorStops == null) {
-      assert(colors.length == 2);
-      gradient.addColorStop(0, colorToCssString(colors[0])!);
-      gradient.addColorStop(1, colorToCssString(colors[1])!);
-      return gradient;
-    } else {
-      for (int i = 0; i < colors.length; i++) {
-        gradient.addColorStop(colorStops[i], colorToCssString(colors[i])!);
-      }
-    }
+    _addColorStopsToCanvasGradient(gradient, colors, colorStops);
     return gradient;
   }
+
+  /// Creates a radial gradient with tiling repeat or mirror.
+  html.CanvasPattern _createGlGradient(html.CanvasRenderingContext2D? ctx,
+      ui.Rect? shaderBounds, double density) {
+    assert(shaderBounds != null);
+    int widthInPixels = shaderBounds!.width.ceil();
+    int heightInPixels = shaderBounds.height.ceil();
+    assert(widthInPixels > 0 && heightInPixels > 0);
+
+    initWebGl();
+    // Render gradient into a bitmap and create a canvas pattern.
+    _OffScreenCanvas offScreenCanvas =
+    _OffScreenCanvas(widthInPixels, heightInPixels);
+    _GlContext gl = _OffScreenCanvas.supported
+        ? _GlContext.fromOffscreenCanvas(offScreenCanvas._canvas!)
+        : _GlContext.fromCanvas(offScreenCanvas._glCanvas!,
+        webGLVersion == WebGLVersion.webgl1);
+    gl.setViewportSize(widthInPixels, heightInPixels);
+
+    NormalizedGradient normalizedGradient = NormalizedGradient(
+        colors, stops: colorStops);
+
+    _GlProgram glProgram = gl.useAndCacheProgram(
+        _WebGlRenderer.writeBaseVertexShader(),
+        _createRadialFragmentShader(normalizedGradient, tileMode))!;
+
+    Object tileOffset = gl.getUniformLocation(glProgram.program, 'u_tile_offset');
+    double centerX = (center.dx - shaderBounds.left) / (shaderBounds.width);
+    double centerY = (center.dy - shaderBounds.top) / (shaderBounds.height);
+    gl.setUniform2f(tileOffset,
+        2 * (shaderBounds.width * (centerX - 0.5)),
+        2 * (shaderBounds.height * (centerY - 0.5)));
+    Object radiusUniform = gl.getUniformLocation(glProgram.program, 'u_radius');
+    gl.setUniform1f(radiusUniform, radius);
+    normalizedGradient.setupUniforms(gl, glProgram);
+    Object gradientMatrix = gl.getUniformLocation(
+        glProgram.program, 'm_gradient');
+    gl.setUniformMatrix4fv(gradientMatrix, false, matrix4 == null ? Matrix4.identity().storage : matrix4!);
```

Comment:
Is identity or null matrix a common case? If so, perhaps we should reuse a static identity matrix instead of instantiating a new one every time. JS-engines are still pretty bad at allocating large typed arrays.

Reply:
Noticed it too. I'm working on a separate CL to go through a lot of other identity() calls we have and optimize identity itself. Matrix4.zero().setIdentity() is not efficient either since Float32List does get already initialized with 0s.